zerver/tests/test_message_delete.py

```diff
@@ -914,7 +915,62 @@ def test_update_first_message_id_on_stream_message_deletion(self) -> None:
         self.assertEqual(stream.first_message_id, message_ids[1])
 
         all_messages = Message.objects.filter(id__in=message_ids)
-        with self.assert_database_query_count(25):
+        with self.assert_database_query_count(26):
             do_delete_messages(realm, all_messages, acting_user=None)
         stream = get_stream(stream_name, realm)
         self.assertEqual(stream.first_message_id, None)
+
+    @mock.patch("zerver.lib.push_notifications.send_push_notifications")
+    @mock.patch("zerver.lib.push_notifications.push_notifications_configured", return_value=True)
+    def test_clear_push_notifications_on_message_deletion(
+        self,
+        mock_push_notifications: mock.MagicMock,
+        mock_send_push_notifications: mock.MagicMock,
+    ) -> None:
+        realm = get_realm("zulip")
+        hamlet = self.example_user("hamlet")
+        cordelia = self.example_user("cordelia")
+        self.register_push_device(cordelia.id)
+        do_change_user_setting(cordelia, "enable_stream_push_notifications", True, acting_user=None)
+
+        def get_message_ids_with_active_push_notification(user_profile: UserProfile) -> list[int]:
+            return list(
+                UserMessage.objects.filter(
+                    user_profile=user_profile,
+                )
+                .extra(  # noqa: S610
+                    where=[UserMessage.where_active_push_notification()],
+                )
+                .order_by("message_id")
+                .values_list("message_id", flat=True)
+            )
+
+        # Initially, no active push notifications.
+        self.assertEqual(get_message_ids_with_active_push_notification(cordelia), [])
+
+        self.subscribe(cordelia, "Verona")
+        message_ids = [self.send_stream_message(hamlet, "Verona", str(i)) for i in range(10)]
```

Comment:
Never write tests that send 10 messages in a loop like this, and doing so makes the test slow. In aggregate, test code like this is very expensive for the project, as the test suite runs a lot. 

I recommend sending 1, or 2 if you need it to exercise a code path, or ~3-5 if you're trying to verify that we're not doing database queries in a loop, and then only if we're actually doing the assert.

Reply:
It looks like we have some existing issues of this form. Perhaps you can do a follow-up project with an audit of `git grep -A5 range zerver/tests/` and fix any existing ones that are sending 5-10 messages in a loop without a reason.